### PR TITLE
[Model] Add Support for Multimodal Granite Models

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR adds the main architectural change needed to support upcoming multimodal granite models from IBM, which are a variant of llava-next. To this end, we need to allow passing a list of `feature_sample_layers`, which are indices in the visual encoder whose hidden states will be concatenated and used as the visual encoder output as an alternative to the last layer. This change should not affect any existing llava next models, and is made in all of the supported visual encoders for Llava-based models in vLLM (clip/siglip/pixtral) to keep things as generic as possible.

These models have not been released quite yet, and we have not yet ported the change to transformers, which is why there are no new tests added. However, I have run my own benchmarks to ensure the results in vLLM are identical in quality to our own implementation, and verified that the `feature_sample_layers` can be passed through each supported visual encoder type as a sanity check. Once the models are out and supported in `transformers`, I can add it to the corresponding test here!

FYI @njhill

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
This looks a bit overcomplicated, and can be easily confused with `num_hidden_layers`. I prefer having a flag to simply return all of the hidden states (similar to HF) and extract them after calling `forward` method.

**Comment by DarkLight1337** [comment]
The reason why we set `num_hidden_layers`  at initialization time is to avoid loading the unused layers. However, there is no such advantage to passing `feature_sample_layers`, so it would be better to just pass the necessary information to `forward`.

**Comment by alex-jw-brooks** [comment]
Sounds good, thanks @DarkLight1337! I moved most of the vision feature layer handling out to inference time, added the flag for getting all of the hidden states + selecting from them after, and fixed the initialization behavior so that we only load up to the deepest layer needed if there are multiple feature layers also

File: vllm/model_executor/models/clip.py [hunk 1]
```diff
@@ -407,12 +416,12 @@ def __init__(
         # NOTE: This typo of "layrnorm" is not fixed on purpose to match
         # the original transformers code and name of the model weights.
         self.pre_layrnorm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
+
         self.encoder = CLIPEncoder(
             config=config,
             quant_config=quant_config,
             num_hidden_layers_override=num_hidden_layers_override,
-            prefix=f"{prefix}.encoder",
```
Comment (DarkLight1337) [inline vllm/model_executor/models/clip.py:414]: Seems like an unnecessary formatting change.

File: vllm/model_executor/models/llava.py [hunk 2]
```diff
@@ -204,7 +204,41 @@ def input_processor_for_llava(ctx: InputContext, inputs: DecoderOnlyInputs):
 
 class LlavaLikeConfig(Protocol):
     vision_config: PretrainedConfig
-    vision_feature_layer: int
+    vision_feature_layer: Union[int, List]
```
Comment (DarkLight1337) [inline vllm/model_executor/models/llava.py:207]:
```suggestion
    vision_feature_layer: Union[int, List[int]]
```

File: vllm/multimodal/utils.py [hunk 3]
```diff
@@ -392,6 +393,47 @@ def encode_video_base64(frames: npt.NDArray):
...
+    offset = max_possible_layers - len(encoder_outputs)
+    hs_pool = [
+        encoder_outputs[layer_idx]
+        if layer_idx >= 0 else encoder_outputs[layer_idx + offset]
+        for layer_idx in feature_sample_layers
+    ]
+
+    # Apply post-norm on the final hidden state if we are using it
+    uses_last_layer = ((len(hs_pool) - 1) in feature_sample_layers
+                       or -1 in feature_sample_layers)
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:431]:
```suggestion
    uses_last_layer = feature_sample_layers[-1] in (len(hs_pool) - 1, -1)
```

File: vllm/multimodal/utils.py [hunk 4]
```diff
@@ -392,6 +393,47 @@ def encode_video_base64(frames: npt.NDArray):
...
+    """
+    if feature_sample_layers is None:
+        if post_layer_norm is not None:
+            return post_layer_norm(encoder_outputs)
+        return encoder_outputs
+
+    # Get the hidden states corresponding to the layer indices.
+    # Negative values are relative to the full visual encoder,
+    # so offset them depending on how many layers were loaded.
+    offset = max_possible_layers - len(encoder_outputs)
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py:422]: This implicitly assumes that encoder outputs include all loaded layers and are in the same order as the layers. Might want to add a note about this.

File: vllm/model_executor/models/clip.py (hunk 1 above)
Comment (alex-jw-brooks) [reply vllm/model_executor/models/clip.py]: Yup! I put it back the way it was, accidentally deleted some trailing commas + auto formatted in some places 😄

File: vllm/multimodal/utils.py (hunk 4 above)
Comment (DarkLight1337) [reply vllm/multimodal/utils.py:425]: I think you made a typo in the comment

**Review by DarkLight1337 — APPROVED** [review]
Otherwise LGTM

**Comment by DarkLight1337** [comment]
cc @ywang96 in case you're planning to refactor related code in the vision encoders.

File: vllm/multimodal/utils.py (hunk 4 above)
Comment (alex-jw-brooks) [reply vllm/multimodal/utils.py:425]: Sure did, thanks for catching that - fixed!
